packages/config/src/Devices.ts

```diff
@@ -41,66 +41,107 @@ export const devicesDir = path.join(configDir, "devices");
+ */
+export async function loadDeviceIndexInternal(
+	forceWrite?: boolean,
+): Promise<DeviceConfigIndex> {
+	// The index file needs to be regenerated if it does not exist
+	let needsUpdate = !(await pathExists(indexPath));
+	let index: DeviceConfigIndex | undefined;
+	let mtimeIndex: Date | undefined;
+	// ...or if cannot be parsed
+	if (!needsUpdate) {
+		try {
+			const fileContents = await readFile(indexPath, "utf8");
+			index = JSON5.parse(fileContents);
+			mtimeIndex = (await fs.stat(indexPath)).mtime;
+		} catch {
+			// console.error("Error while parsing index file - regenerating...");
+			needsUpdate = true;
+		} finally {
+			if (!index) {
+				// console.error("Index file was malformed - regenerating...");
+				needsUpdate = true;
+			}
+		}
+	}
 
-	const index: DeviceConfigIndex = [];
-
-	for (const file of configFiles) {
-		const relativePath = path
-			.relative(devicesDir, file)
-			.replace(/\\/g, "/");
-		const fileContents = await readFile(file, "utf8");
-		// Try parsing the file
-		const config = new DeviceConfig(relativePath, fileContents);
-		// Add the file to the index
-		index.push(
-			...config.devices.map((dev: any) => ({
-				manufacturerId: formatId(config.manufacturerId.toString(16)),
-				...dev,
-				firmwareVersion: config.firmwareVersion,
-				filename: relativePath,
-			})),
-		);
+	// ...or if there were any changes in the file system
+	if (!needsUpdate) {
+		needsUpdate = await hasChangedDeviceFiles(devicesDir, mtimeIndex!);
 	}
 
-	if (!!process.env.CI || process.env.NODE_ENV !== "test") {
-		// Write the index (but not during tests)
-		await writeFile(
-			path.join(devicesDir, "index.json"),
-			`// This file is auto-generated. DO NOT edit it by hand if you don't know what you're doing!"
-	${stringify(index, "\t")}
-	`,
-			"utf8",
+	if (needsUpdate) {
+		index = [];
+
+		const configFiles = await enumFilesRecursive(
+			devicesDir,
+			(file) => file.endsWith(".json") && !file.endsWith("index.json"),
 		);
-	}
 
-	return index;
-}
+		for (const file of configFiles) {
+			const relativePath = path
+				.relative(devicesDir, file)
+				.replace(/\\/g, "/");
+			const fileContents = await readFile(file, "utf8");
+			// Try parsing the file
+			const config = new DeviceConfig(relativePath, fileContents);
+			// Add the file to the index
+			index.push(
+				...config.devices.map((dev: any) => ({
+					manufacturerId: formatId(
+						config.manufacturerId.toString(16),
+					),
+					...dev,
+					firmwareVersion: config.firmwareVersion,
+					filename: relativePath,
+				})),
+			);
+		}
+
+		if (forceWrite || process.env.NODE_ENV !== "test") {
+			// Save the index to disk (but not during unit tests)
+			await writeFile(
+				path.join(devicesDir, "index.json"),
+				`// This file is auto-generated. DO NOT edit it by hand if you don't know what you're doing!"
+${stringify(index, "\t")}
+`,
+				"utf8",
+			);
+			console.log(`Device index regenerated`);
```

Comment:
Use logger here

Reply:
It is not guaranteed that we have a logger instance available. The code is also used in unit and CI tests without a driver instance.

Reply:
So log if logger is present and use console if not?

Reply:
I'm not so invested in that log line, I'll rather remove it altogether 😅
Too much hassle passing the driver instance through ;)

Reply:
I mean it's useful so for example if log to file is enabled you will see that